pkg/collector/corechecks/cluster/orchestrator/orchestrator_test.go

```diff
@@ -0,0 +1,98 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver && orchestrator
+
+package orchestrator
+
+import (
+	"context"
+	"sync"
+	"testing"
+
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/cluster/orchestrator/collectors"
+	"github.com/DataDog/datadog-agent/pkg/collector/corechecks/cluster/orchestrator/collectors/inventory"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/apiserver"
+
+	"github.com/stretchr/testify/assert"
+	corev1 "k8s.io/api/core/v1"
+	crd "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	vpa "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/client/clientset/versioned/fake"
+	"k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+)
+
+func newCollectorBundle(chk *OrchestratorCheck) *CollectorBundle {
+	bundle := &CollectorBundle{
+		discoverCollectors: chk.orchestratorConfig.CollectorDiscoveryEnabled,
+		check:              chk,
+		inventory:          inventory.NewCollectorInventory(),
+		runCfg: &collectors.CollectorRunConfig{
+			APIClient:                   chk.apiClient,
+			ClusterID:                   chk.clusterID,
+			Config:                      chk.orchestratorConfig,
+			MsgGroupRef:                 chk.groupID,
+			OrchestratorInformerFactory: chk.orchestratorInformerFactory,
+		},
+		stopCh:              chk.stopCh,
+		manifestBuffer:      NewManifestBuffer(chk),
+		activatedCollectors: map[string]struct{}{},
+	}
+	bundle.importCollectorsFromInventory()
+	bundle.prepareExtraSyncTimeout()
+	return bundle
+}
+
+// TestOrchestratorCheckSafeReSchedule close simulates the check being unscheduled and rescheduled again
+func TestOrchestratorCheckSafeReSchedule(t *testing.T) {
+	var wg sync.WaitGroup
+
+	client := fake.NewSimpleClientset()
+	vpaClient := vpa.NewSimpleClientset()
+	crdClient := crd.NewSimpleClientset()
+	cl := &apiserver.APIClient{Cl: client, VPAClient: vpaClient, CRDClient: crdClient}
+	orchCheck := OrchestratorFactory().(*OrchestratorCheck)
+	orchCheck.apiClient = cl
+
+	orchCheck.orchestratorInformerFactory = getOrchestratorInformerFactory(cl)
+	bundle := newCollectorBundle(orchCheck)
+	err := bundle.Initialize()
+	assert.NoError(t, err)
+
+	wg.Add(2)
+
+	// getting rescheduled.
+	orchCheck.Cancel()
+
+	bundle.runCfg.OrchestratorInformerFactory = getOrchestratorInformerFactory(cl)
+	bundle.stopCh = make(chan struct{})
+	err = bundle.Initialize()
+	assert.NoError(t, err)
+
+	bundle.runCfg.OrchestratorInformerFactory.InformerFactory.Core().V1().Nodes().Informer().AddEventHandler(&cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			wg.Done()
+		},
+	})
+
+	writeNode(t, client, "1")
+	writeNode(t, client, "2")
+
+	wg.Wait()
```

Comment:
is it worth adding a time out here? I think this test will never fail - just run forever.

Reply:
https://github.com/DataDog/datadog-agent/pull/15616/commits/45b10cbeb819685f35c18d91e67a2490a8076d66